src/Analyzers/Core/CodeFixes/NamespaceMatchFolder/AbstractChangeNamespaceToMatchFolderCodeFixProvider.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Analyzers.NamespaceSync;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Rename;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeFixes.NamespaceMatchFolder
+{
+    internal abstract partial class AbstractChangeNamespaceToMatchFolderCodeFixProvider : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                new MyCodeAction(AnalyzersResources.Change_namespace_to_match_folder_structure, cancellationToken => FixAllInDocumentAsync(context.Document, context.Diagnostics, cancellationToken)),
+                context.Diagnostics);
+
+            return Task.CompletedTask;
+        }
+
+        private static async Task<Solution> FixAllInDocumentAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            // All the target namespaces should be the same for a given document
+            Debug.Assert(diagnostics.Select(diagnostic => diagnostic.Properties[AbstractNamespaceMatchFolderDiagnosticAnalyzer.TargetNamespace]).Distinct().Count() == 1);
+
+            var targetNamespace = diagnostics.First().Properties[AbstractNamespaceMatchFolderDiagnosticAnalyzer.TargetNamespace];
+            RoslynDebug.AssertNotNull(targetNamespace);
+
+            // Use the Renamer.RenameDocumentAsync API to sync namespaces in the document. This allows
+            // us to keep in line with the sync methodology that we have as a public API and not have 
+            // to rewrite or move the complex logic. RenameDocumentAsync is designed to behave the same
+            // as the intent of this analyzer/codefix pair.
+            var targetFolders = PathMetadataUtilities.BuildFoldersFromNamespace(targetNamespace);
+            var documentWithNoFolders = document.WithFolders(Array.Empty<string>());
+            var renameActionSet = await Renamer.RenameDocumentAsync(
+                documentWithNoFolders,
+                documentWithNoFolders.Name,
+                newDocumentFolders: targetFolders,
+                cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
i'm really confused here.  this seems to be updating the folders for this document.  i.e. you're setting 'targetFolders'.  But the feature name is ChangeNamespaceToMatchFolder  not ChangeFolderToMatchNamespace...

why are the folders changing here as opposed to just chnaging the namespace?

Reply:
I really did my best to explain this above. This is a convenient way to update the namespace in a document using a public API we have without having to expose more internals. It has the same requirements and fits in the same featureset. The only difference is the oddity of forcing the folder update in order to use the API.